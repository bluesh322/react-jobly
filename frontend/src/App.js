import React, { useState, useEffect } from "react";
import useLocalStorage from "./hooks/useLocalStorage";
import Routes from "./routes-nav/Routes";
import { BrowserRouter } from "react-router-dom";
import NavBar from "./routes-nav/Navbar";
import JoblyApi from "./api";
import UserContext from "./auth/UserContext";
import jwt from "jsonwebtoken";

// Key name for storing JWT token in local storage
export const TOKEN_STORAGE_ID = "jobly-token";

/**
 * Jobly application
 * 
 * -infoLoaded: has user info been loaded? (manages "loading ...")
 * 
 * -currentUser: user obj from API. Our way to see if a user is logged in,
 *  passed around via context.
 * 
 * -token: for logged in users, this is auth JWT.
 *  Is required for most API calls, and is initially read from LocalStorage
 *  and synced to there via the useLocalStorage hook.
 * 
 * App -> Routes
 */

const App = () => {
  const [infoLoaded, setInfoLoaded] = useState(false);
  const [currentUser, setCurrentUser] = useState(null);
  const [token, setToken] = useLocalStorage(TOKEN_STORAGE_ID);
  const [applicationIds, setApplicationIds] = useState(new Set([]));

  /** Load user info from API. Until a user has a token it should not run.
   *  Will only rerun if a user logs out.
   */

  useEffect(() => {
    const getCurrentUser = async () => {
      if (token) {
        try {
          let { username } = jwt.decode(token);
          JoblyApi.token = token;
          let currentUser = await JoblyApi.getCurrentUser(username);
          setCurrentUser(currentUser);
          //setAppIDs
        } catch (errors) {
          console.error("App loadUserInfo: problem loading", errors);
          setCurrentUser(null);
        }
      }
      setInfoLoaded(true);
    }
    setInfoLoaded(false);
    getCurrentUser();
  }, [token])

  /** Signup
   *
   * Automatically logs them in upon signup.
   */

  const signup = async (signupData) => {
    try {
      let token = await JoblyApi.signup(signupData);
      setToken(token);
      return { success: true };
    } catch (errors) {
      console.log("signup failed", errors);
      return { success: false, errors };
    }
  };

  /** Login
   * 
   * 
   */
  const login = async (loginData) => {
    try {
      let token = await JoblyApi.login(loginData);
      setToken(token);
      return { success: true };
    } catch (errors) {
      console.log("login failed", errors);
      return { success: false, errors };
    }
  }

  const logout = async () => {
    setCurrentUser(null);
    setToken(null);
  }

  const applyForJob = (id) => {
    if(hasAppliedForJob(id)) return;
    JoblyApi.applyForJob(currentUser.username, id);
    setApplicationIds(new Set([...applicationIds, id]));
  }

  const hasAppliedForJob = (id) => {
    return applicationIds.has(id);
  }

  if (!infoLoaded) return <span data-testid="loading">Loading ...</span>;

  return (
    <BrowserRouter>
      <UserContext.Provider
        value={{ currentUser, setCurrentUser, hasAppliedForJob, applyForJob}}
      >
        <div className="App">
          <NavBar logout={logout} />
          <Routes login={login} signup={signup}/>
        </div>
      </UserContext.Provider>
    </BrowserRouter>
  );
};

export default App;
