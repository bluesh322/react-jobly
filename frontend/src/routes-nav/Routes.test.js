import { render, fireEvent, act } from "@testing-library/react";
import { MemoryRouter } from "react-router";
import App from "../App";
import JoblyApi from "../api";

jest.mock("../api");

const token =
  "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJ1c2VybmFtZSI6InRlc3QiLCJpc0FkbWluIjpmYWxzZSwiaWF0IjoxNjE4MzQ5MTc1fQ.7gzrL0AmPZP9rmm7rR3bd5ZMS9DQIQosj_qwmfTn170";

const testUser = {
  username: "testuser",
  firstName: "test",
  lastName: "test",
  email: "t@t.com",
  isAdmin: "false",
  applications: [],
};

const testCompany = {
  handle: "anderson-aria-morrow",
  name: "Anderson, Arias and Morrow",
  description: "Somebody program how I.",
  numEmployees: 245,
  logoUrl: "/logos/logo3.png",
};

const testJob = {
  id: 7,
  title: "Technical brewer",
  salary: 157000,
  equity: "0",
  companyHandle: "anderson-aria-morrow",
};

beforeEach(async () => {
  await JoblyApi.login.mockResolvedValue(token);
  await JoblyApi.getCurrentUser.mockResolvedValue(testUser);
  await JoblyApi.getJobs.mockResolvedValue([testJob]);
  await JoblyApi.applyForJob.mockResolvedValue(testJob.id);
  await JoblyApi.getCompanies.mockResolvedValue([testCompany]);
  await JoblyApi.getCompany.mockResolvedValue({
    ...testCompany,
    jobs: [testJob],
  });
});

afterEach(() => {
  jest.resetAllMocks();
});

describe("logged out views", () => {
  test("renders without crashing", () => {
    render(
      <MemoryRouter>
        <App />
      </MemoryRouter>
    );
  });

  it("matches snapshot", () => {
    const { asFragment } = render(
      <MemoryRouter>
        <App />
      </MemoryRouter>
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it("displays homepage", async () => {
    const { getByText } = render(
      <MemoryRouter initialEntries={["/"]}>
        <App />
      </MemoryRouter>
    );

    expect(getByText("Jobly")).toBeInTheDocument();
  });

  it("redirects to protected route if not logged in", async () => {
    const { getByText } = render(
      <MemoryRouter initialEntries={["/jobs"]}>
        <App />
      </MemoryRouter>
    );

    expect(getByText("Jobly")).toBeInTheDocument();
  });

  it("displays signup form when nav link is clicked", async () => {
    const { getAllByText, getByLabelText } = render(
      <MemoryRouter initialEntries={["/"]}>
        <App />
      </MemoryRouter>
    );

    await act(async () => {
      fireEvent.click(getAllByText("Sign Up")[0]);
    });

    expect(getByLabelText("password")).toBeInTheDocument();
  });
});

describe("logged in views", () => {
  it("let a user log in ", async () => {
    const { getAllByText, getByLabelText, findByText, getByTestId } = render(
      <MemoryRouter initialEntries={["/"]}>
        <App />
      </MemoryRouter>
    );

    await act(async () => {
      fireEvent.click(getAllByText("Login")[0]);
    });

    expect(getByLabelText("username")).toBeInTheDocument();

    fireEvent.input(getByLabelText("username"), {
      target: { value: "test" },
    });

    fireEvent.input(getByLabelText("password"), {
      target: { value: "password" },
    });

    fireEvent.submit(getByTestId("submit"));
    expect(await findByText("Welcome Back testuser")).toBeInTheDocument();

    //I HAVE TO MANUALLY LOG IN AND BACK OUT EACH TEST
    await act(async () => {
      fireEvent.click(getAllByText("Log Out")[0]);
    });
  });

  it("let a user log in and view /companies", async () => {
    const {
      getByText,
      getAllByText,
      getByLabelText,
      findByText,
      getByTestId,
    } = render(
      <MemoryRouter initialEntries={["/"]}>
        <App />
      </MemoryRouter>
    );

    await act(async () => {
      fireEvent.click(getAllByText("Login")[0]);
    });

    expect(getByLabelText("username")).toBeInTheDocument();

    fireEvent.input(getByLabelText("username"), {
      target: { value: "test" },
    });

    fireEvent.input(getByLabelText("password"), {
      target: { value: "password" },
    });

    fireEvent.submit(getByTestId("submit"));
    expect(await findByText("Welcome Back testuser")).toBeInTheDocument();

    await act(async () => {
      fireEvent.click(getByText("Companies"));
    });

    let company = await findByText("Anderson, Arias and Morrow");
    expect(company).toBeInTheDocument();

    await act(async () => {
      fireEvent.click(company);
    });
    expect(await findByText("Technical brewer")).toBeInTheDocument();

    await act(async () => {
      fireEvent.click(getAllByText("Log Out")[0]);
    });
  });

  it("let a user log in and view /jobs and apply to a job", async () => {
    const {
      getByText,
      getAllByText,
      getByLabelText,
      findByText,
      getByTestId,
    } = render(
      <MemoryRouter initialEntries={["/"]}>
        <App />
      </MemoryRouter>
    );

    await act(async () => {
      fireEvent.click(getAllByText("Login")[0]);
    });

    expect(getByLabelText("username")).toBeInTheDocument();

    fireEvent.input(getByLabelText("username"), {
      target: { value: "test" },
    });

    fireEvent.input(getByLabelText("password"), {
      target: { value: "password" },
    });

    fireEvent.submit(getByTestId("submit"));
    expect(await findByText("Welcome Back testuser")).toBeInTheDocument();

    await act(async () => {
      fireEvent.click(getByText("Jobs"));
    });

    let company = await findByText("Technical brewer");
    expect(company).toBeInTheDocument();

    await act(async () => {
      fireEvent.click(getByTestId("apply"));
    });
    expect(await findByText("Technical brewer")).toBeInTheDocument();

    expect(await findByText("Applied")).toBeInTheDocument();

    await act(async () => {
      fireEvent.click(getAllByText("Log Out")[0]);
    });
  });

  it("let a user log in and log out", async () => {
    const {
      queryByText,
      getAllByText,
      getByLabelText,
      findByText,
      getByTestId,
    } = render(
      <MemoryRouter initialEntries={["/"]}>
        <App />
      </MemoryRouter>
    );

    await act(async () => {
      fireEvent.click(getAllByText("Login")[0]);
    });

    expect(getByLabelText("username")).toBeInTheDocument();

    fireEvent.input(getByLabelText("username"), {
      target: { value: "test" },
    });

    fireEvent.input(getByLabelText("password"), {
      target: { value: "password" },
    });

    fireEvent.submit(getByTestId("submit"));
    expect(await findByText("Welcome Back testuser")).toBeInTheDocument();

    await act(async () => {
      fireEvent.click(getAllByText("Log Out")[0]);
    });

    expect(await queryByText("Welcome Back testuser")).not.toBeInTheDocument();
  });
});
