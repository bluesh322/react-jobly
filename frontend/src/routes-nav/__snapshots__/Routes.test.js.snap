// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`logged out views matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="App"
  >
    <header
      class="MuiPaper-root MuiAppBar-root MuiAppBar-positionStatic MuiAppBar-colorPrimary MuiPaper-elevation4"
    >
      <div
        class="MuiToolbar-root MuiToolbar-regular MuiToolbar-gutters"
      >
        <div
          class="MuiContainer-root makeStyles-navDisplayFlex-4 MuiContainer-maxWidthMd"
        >
          <button
            aria-label="home"
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-edgeStart"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiIconButton-label"
            >
              <a
                href="/"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeLarge"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M10 20v-6h4v6h5v-8h3L12 3 2 12h3v8z"
                  />
                </svg>
              </a>
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
          <nav
            aria-labelledby="main navigation"
            class="MuiList-root makeStyles-navDisplayFlex-4 MuiList-padding"
          >
            <a
              class="makeStyles-linkText-5"
              href="/signup"
            >
              <div
                aria-disabled="false"
                class="MuiButtonBase-root MuiListItem-root MuiListItem-gutters MuiListItem-button"
                role="button"
                tabindex="0"
              >
                <div
                  class="MuiListItemText-root"
                >
                  <span
                    class="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
                  >
                    Sign Up
                  </span>
                </div>
                <span
                  class="MuiTouchRipple-root"
                />
              </div>
            </a>
            <a
              class="makeStyles-linkText-5"
              href="/login"
            >
              <div
                aria-disabled="false"
                class="MuiButtonBase-root MuiListItem-root MuiListItem-gutters MuiListItem-button"
                role="button"
                tabindex="0"
              >
                <div
                  class="MuiListItemText-root"
                >
                  <span
                    class="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
                  >
                    Login
                  </span>
                </div>
                <span
                  class="MuiTouchRipple-root"
                />
              </div>
            </a>
          </nav>
        </div>
      </div>
    </header>
    <div
      class="MuiContainer-root MuiContainer-maxWidthMd"
    >
      <h2
        data-testid="resolved"
      >
        Jobly
      </h2>
      <p>
        All the jobs in one, convenient place
      </p>
      <div
        class="MuiBox-root MuiBox-root-6"
      >
        <a
          href="/signup"
        >
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiButton-label"
            >
              Sign Up
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </a>
        <a
          href="/login"
        >
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedSecondary"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiButton-label"
            >
              Login
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </a>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
